subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultVisualStudioLocator.java

```diff
@@ -105,105 +100,149 @@ public SearchResult locateDefaultVisualStudioInstall(File candidate) {
-            locateInstallsInRegistry(baseKey);
+    private void locateInstallsWith(VisualStudioVersionLocator versionLocator) {
+        List<VisualStudioMetadata> installs = versionLocator.getVisualStudioInstalls();
+
+        for (VisualStudioMetadata install : installs) {
+            addInstallIfValid(install);
         }
     }
 
-    private void locateInstallsInRegistry(String baseKey) {
-        List<String> visualCppVersions;
-        try {
-            visualCppVersions = windowsRegistry.getValueNames(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, baseKey + REGISTRY_ROOTPATH_VC);
-        } catch (MissingRegistryEntryException e) {
-            // No Visual Studio information available in the registry
-            return;
-        }
+    private void addInstallIfValid(VisualStudioMetadata install) {
+        File visualCppDir = install.getVisualCppDir();
+        File visualStudioDir = install.getInstallDir();
 
-        for (String valueName : visualCppVersions) {
-            if (!valueName.matches("\\d+\\.\\d+")) {
-                // Ignore the other values
-                continue;
-            }
-            File visualCppDir = new File(windowsRegistry.getStringValue(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, baseKey + REGISTRY_ROOTPATH_VC, valueName));
-            visualCppDir = FileUtils.canonicalize(visualCppDir);
-            File visualStudioDir = visualCppDir.getParentFile();
-
-            if (isVisualCpp(visualCppDir) && isVisualStudio(visualStudioDir)) {
-                LOGGER.debug("Found Visual C++ {} at {}", valueName, visualCppDir);
-                VersionNumber version = VersionNumber.parse(valueName);
-                VisualCppInstall visualCpp = buildVisualCppInstall("Visual C++ " + valueName, visualStudioDir, visualCppDir, version);
-                VisualStudioInstall visualStudio = new VisualStudioInstall(visualStudioDir, visualCpp);
-                foundInstalls.put(visualStudioDir, visualStudio);
-            } else {
-                LOGGER.debug("Ignoring candidate Visual C++ directory {} as it does not look like a Visual C++ installation.", visualCppDir);
-            }
+        if (isValidInstall(install)) {
+            LOGGER.debug("Found Visual C++ {} at {}", install.getVersion(), visualCppDir);
+            VersionNumber version = install.getVersion();
+            VisualCppInstall visualCpp = buildVisualCppInstall("Visual C++ " + install.getVisualCppVersion(), visualStudioDir, visualCppDir, install.getVisualCppVersion(), install.getCompatibility());
+            VisualStudioInstall visualStudio = new VisualStudioInstall("Visual Studio " + version, visualStudioDir, version, visualCpp);
+            foundInstalls.put(visualStudioDir, visualStudio);
+        } else {
+            LOGGER.debug("Ignoring candidate Visual C++ directory {} as it does not look like a Visual C++ installation.", visualCppDir);
         }
     }
 
     private void locateInstallInPath() {
```

Comment:
feels like this should live in another VS install locator.

Reply:
Yeah, that was my initial plan, but then it just turned into a thin wrapper around `OperationSystem.findInPath()` which felt like overkill.  I suppose it still makes sense to break it out, though.

Reply:
This has been moved to its own version locator.